src/Compiler/Optimize/LowerComputedCollections.fs

```diff
@@ -267,15 +267,24 @@ module List =
         let mkListInit mkLoop =
             mkCompGenLetMutableIn m "collector" collectorTy (mkDefault (m, collectorTy)) (fun (_, collector) ->
                 let reader = InfoReader (g, amap)
-                let loop = mkLoop (fun _idxVar loopVar -> mkCallCollectorAdd tcVal g reader m collector loopVar)
+
+                let loop =
+                    mkLoop (fun _idxVar loopVar ->
+                        let body =
+                            body
+                            |> Option.map (fun (loopVal, body) -> mkInvisibleLet m loopVal loopVar body)
+                            |> Option.defaultValue loopVar
+
+                        mkCallCollectorAdd tcVal g reader m collector body)
```

Comment:
It might be possible to avoid doing this binding in the body of the loop, but:

- Maybe the JIT can optimize it away anyway?
- If it has any detrimental effect on perf, it's dwarfed by the overall benefit of applying the optimization as a whole.
- It seems like it would require more significant refactoring, which, given the above, doesn't seem likely to be worth the effort (to me).